### ISSUE TITLE
The bad generations of `generate.py`.

### ISSUE THREAD

**Issue by louis2889184** [lucidrains/DALLE-pytorch#120]
Thanks for the repo.

I trained the DALLE on visual genome dataset. During the training, one of the generations is shown below,

![image](https://user-images.githubusercontent.com/32589903/112347302-19784a80-8c9d-11eb-8635-e388752e24c8.png)
![image](https://user-images.githubusercontent.com/32589903/112347338-2301b280-8c9d-11eb-9ec9-7f58c2ab9425.png)

But when I want to generate an image by `generate.py`, the generated images are non-sense, even though I use the text also appeared in the training phase.

The scripts I use
```
# separate the punctuation marks with characters
python generate.py --dalle_path ./dalle.pt --text "tire on bus . window on bus . window on bus . window on bus . window on bus . pole in grass . window on bus .
```
and 
```
# don't separate the punctuation marks with characters
python generate.py --dalle_path ./dalle.pt --text "tire on bus. window on bus. window on bus. window on bus. window on bus. pole in grass. window on bus."
```

The results of both scripts are similar, and the generations are

![image](https://user-images.githubusercontent.com/32589903/112348477-3d885b80-8c9e-11eb-8637-7c01e4efd52a.png)
![image](https://user-images.githubusercontent.com/32589903/112348487-40834c00-8c9e-11eb-84cf-61ae7d1de53f.png)
![image](https://user-images.githubusercontent.com/32589903/112348498-4416d300-8c9e-11eb-89a4-c357010f7977.png)

I have checked the model weights are loaded normally. Any thoughts on this issue?

**Comment by robvanvolt** [comment]
The training datasets have to be drastically increased to get decent results - have a look a few issues underneath on the results of afiaka87: https://github.com/lucidrains/DALLE-pytorch/issues/86#issue-832121328

**Comment by louis2889184** [comment]
I see. Maybe I didn't describe the issue clearly. My problem is more about the mismatched results between images generated during training and by `generate.py`. As you can see, the two outputs look very different even I use the same input text. Moreover, no matter what I use as the input text, the generated images look similar, which contains lots of black blocks.

**Comment by afiaka87** [comment]
> I see. Maybe I didn't describe the issue clearly. My problem is more about the mismatched results between images generated during training and by `generate.py`. As you can see, the two outputs look very different even I use the same input text. Moreover, no matter what I use as the input text, the generated images look similar, which contains lots of black blocks.

@louis2889184 

I have noticed similar behavior with reconstructions (even ones directly from the training set) tend to be quite blurry and abstract. I'm out of my depth on that - but I assume it is due to the transformer being forced to recreate the phrase verbatim while allowing image output to deviate a bit. Hopefully someone more knowledgeable than I can chime in on why this happens.

As for the "mismatch" between images generated:

a.) it wont be deterministic. Run generation many times and see if you can find one or two that are similar.
b.) if your model hasn't generalized (and it needs a kind of absurd amount of data to do so) then even the _slightest_ deviation in your phrasing compared to what is in the training will give you totally different output. Is the phrase _identical_?
c.) Most importantly: what are your parameters? Learning rate, batch_size, depth, heads? These numbers dramatically effect the output.

**Comment by louis2889184** [comment]
@afiaka87 Thanks for your help. I have three different texts corresponding to the same image, and will randomly pick one when loading the data. I'll try to make the image and text one to one, and see if the mismatch will go away.

**Comment by afiaka87** [comment]
@lucidrains I haven't yet reproduced this exact issue, but I have discussed it extensively on the discord and am at least having trouble understanding why @louis2889184 generations are in fact decidedly worse than the reconstructions from training even though they're providing the exact same text. Can you clarify on why that might happen? To be clear - I've only used the generate code a few times.

**Comment by afiaka87** [comment]
@louis2889184 let me know here once you have tried re-running with lower depth/heads and reversible turned off.

**Comment by louis2889184** [comment]
@afiaka87 It turns out it's because the approach we save images is different in `train_dalle.py` and `generate.py`. 

In `train_dalle.py` we use `wandb.Image` to process the image, and it will automatically normalize and scale the image.
https://github.com/wandb/client/blob/9cc04578ebc6d593450e9dbbcae07452bf7bec35/wandb/sdk/data_types.py#L1676-L1679

However, in `generate.py`, we use `torchvision.utils.save_image` to do it. It won't normalize the image to (0, 1) unless we specify the argument `normalize=True`. Also, the VAE's output range is roughly within -1 and 1, if we don't normalize the image, 
`save_image` will directly transform the float number to uint8 by
```
ndarr = grid.mul(255).add_(0.5).clamp_(0, 255).permute(1, 2, 0).to('cpu', torch.uint8).numpy()
```
Hence, there will be lots of pixels, which are original smaller than 0, become 0. That's why there is a big part of the image is black.

There are some results, the config I use is
```
EPOCHS = 20
BATCH_SIZE = 8
LEARNING_RATE = 3e-4
GRAD_CLIP_NORM = 0.5

MODEL_DIM = 256 # 512
TEXT_SEQ_LEN = 64 # 256
DEPTH = 32
HEADS = 16
DIM_HEAD = 64
REVERSIBLE = False
ATTN_TYPES = None
```

And the outputs of `generate.py`, given text "frame on wall. lamp by bed. wall on building.", are
![0](https://user-images.githubusercontent.com/32589903/112733379-20f65880-8f16-11eb-9b02-820728efebdd.jpg)
![1](https://user-images.githubusercontent.com/32589903/112733380-20f65880-8f16-11eb-8c7b-8fa85fec92a5.jpg)
![2](https://user-images.githubusercontent.com/32589903/112733381-20f65880-8f16-11eb-8fca-bf45083b7aee.jpg)
![3](https://user-images.githubusercontent.com/32589903/112733382-20f65880-8f16-11eb-92cc-c3887571fa17.jpg)
![4](https://user-images.githubusercontent.com/32589903/112733383-218eef00-8f16-11eb-801b-847dfcecdf6f.jpg)

After I add `normalize=True` in `save_image(image, outputs_dir / f'{i}.jpg', normalize=True)`, the outputs are
![0](https://user-images.githubusercontent.com/32589903/112733430-66b32100-8f16-11eb-83bc-369c4d1072e4.jpg)
![1](https://user-images.githubusercontent.com/32589903/112733431-66b32100-8f16-11eb-8084-63d94169b531.jpg)
![2](https://user-images.githubusercontent.com/32589903/112733433-66b32100-8f16-11eb-9970-336e9c10649f.jpg)
![3](https://user-images.githubusercontent.com/32589903/112733434-674bb780-8f16-11eb-9dfa-c8720d08a093.jpg)
![4](https://user-images.githubusercontent.com/32589903/112733435-674bb780-8f16-11eb-88c0-8e710f5191cc.jpg)

Looks much better.

BTW, the `mask` seems also important for generations. `output = dalle.generate_images(text_chunk, mask = mask, filter_thres = args.top_k)`. The above results are generated by inputting mask. But I haven't dig into this too much.

**Comment by afiaka87** [comment]
This is great work and an obvious opportunity to submit a pull request if you'd like. @louis2889184

**Comment by afiaka87** [comment]
A simple inline edit with the github ui should make this a fairly easy fix.

**Comment by louis2889184** [comment]
The PR is made! Thanks, @afiaka87.
